targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/api/lp_ticker.c

```diff
@@ -71,102 +93,113 @@ void lp_ticker_init(void)
+        }
 
-    /* Setup low resolution clock - RTC */
-    if (!rtc_isenabled()) {
-        rtc_init();
+        RTC->TAR = 0; /* Write clears the IRQ flag */
+        NVIC_ClearPendingIRQ(RTC_IRQn);
         RTC_DisableInterrupts(RTC, kRTC_AlarmInterruptEnable | kRTC_SecondsInterruptEnable);
-        RTC_StartTimer(RTC);
-    }
+        NVIC_SetVector(RTC_IRQn, (uint32_t) rtc_isr);
+        NVIC_EnableIRQ(RTC_IRQn);
 
-    RTC->TAR = 0; /* Write clears the IRQ flag */
-    NVIC_ClearPendingIRQ(RTC_IRQn);
-    NVIC_SetVector(RTC_IRQn, (uint32_t)rtc_isr);
-    NVIC_EnableIRQ(RTC_IRQn);
-
-    /* Setup high resolution clock - LPTMR */
-    LPTMR_GetDefaultConfig(&lptmrConfig);
-    /* Use 32kHz drive */
-    CLOCK_SetXtal32Freq(OSC32K_CLK_HZ);
-    lptmrConfig.prescalerClockSource = kLPTMR_PrescalerClock_2;
-    LPTMR_Init(LPTMR0, &lptmrConfig);
-    LPTMR_EnableInterrupts(LPTMR0, kLPTMR_TimerInterruptEnable);
-    NVIC_ClearPendingIRQ(LPTMR0_IRQn);
-    NVIC_SetVector(LPTMR0_IRQn, (uint32_t)lptmr_isr);
-    EnableIRQ(LPTMR0_IRQn);
+        /* Setup high resolution clock - LPTMR */
+        LPTMR_GetDefaultConfig(&lptmrConfig);
+
+        /* Use 32kHz drive */
+        CLOCK_SetXtal32Freq(OSC32K_CLK_HZ);
+        lptmrConfig.prescalerClockSource = kLPTMR_PrescalerClock_2;
+        LPTMR_Init(LPTMR0, &lptmrConfig);
+        LPTMR_DisableInterrupts(LPTMR0, kLPTMR_TimerInterruptEnable);
+        NVIC_ClearPendingIRQ(LPTMR0_IRQn);
+        NVIC_SetVector(LPTMR0_IRQn, (uint32_t) lptmr_isr);
+        EnableIRQ(LPTMR0_IRQn);
+
+        lptmr_schedule = 0;
+
+        lp_ticker_inited = true;
+    } else {
+        /* In case of re-init we need to disable lp ticker interrupt. */
+        LPTMR_StopTimer(LPTMR0);
+
+        RTC_DisableInterrupts(RTC, kRTC_AlarmInterruptEnable);
+        RTC->TAR = 0; /* Write clears the IRQ flag */
+
+        lptmr_schedule = 0;
+    }
 }
 
 /** Read the current counter
  *
- * @return The current timer's counter value in microseconds
+ * @return The current timer's counter value in ticks
  */
 uint32_t lp_ticker_read(void)
 {
-    uint32_t sec, pre;
-
-    if (!lp_ticker_inited) {
-        lp_ticker_init();
-    }
-
-    sec = RTC->TSR; /* 32b: Seconds */
-    pre = RTC->TPR; /* 16b: Increments every 32.768kHz clock cycle (30us) */
-
-    /* Final value: 11b (4095) for sec and 21b for usec (pre can reach 1,000,000us which is close to 1<<20) */
-    uint32_t ret = (((sec & MAX_SEC_MASK) * SEC_IN_USEC) + (((uint64_t)pre * SEC_IN_USEC) / OSC32K_CLK_HZ));
-
-    return ret;
+    /* TPR is increments every 32.768 kHz clock cycle. The TSR increments when
+     * bit 14 of the TPR transitions from a logic one (32768 ticks - 1 sec).
+     * After that TPR starts counting from 0.
+     *
+     * count value is built as follows:
+     * count[0 - 14] - ticks (RTC->TPR)
+     * count[15 - 31] - seconds (RTC->TSR)
+     */
+    const uint32_t count = ((RTC->TSR << SEC_SHIFT) | (RTC->TPR & TICKS_MASK));
```

Comment:
The Kinetis RTC is based on a ripple counter so if it is read at the right time it can glich. To guarantee a glitch free value, the same count needs to be read twice in a row. A psuedo code example of this is here:
https://github.com/ARMmbed/mbed-os/blob/2f14057b5c7ae2e02f5ceeac8e04548b1e143d21/hal/lp_ticker_api.h#L143-L149

This would be a good case to add a test for if we don't have one already. Repeatedly call `lp_ticker_read` in a tight loop and if the value every glitches into the past then fail the test.

Reply:
I have fixed `lp_ticker_read` as suggested and added glitch test cases for lp ticker.